UriLabelService.getWorkspaceLabel

isidorn opened the pull request:
This PR does the following:
* Moves the `UriLabelService` to be a core service, such that the configuration service depends on it, not the other way around. Note that the previous solution was also unhappy with the `configurationService.acquireUriLabelService`. Now it is the other way around. The reason for this is that the `configurationService` needs the `workspaceLabel` very early on startup
* Adds the `getWorkspaceLabel()` to the `UriLabelService`
* Adopts the `getWorkspaceLabel` acrross the workbench and removes the weird function from `workspaces.ts`

The benefits of this is that we can easily add rules for computing a workspace label. For example instead of the `remotehub` scheme we could display `github`.

I still plan to simplify the implementation of the `getWorkspaceLabel` and adopt it to show proper label for werid workspace schemas.
What might be an issue: in some cases the workspace label is needed very early and this might be before clients get a chance to register "rules" for displaying the workspace label - though will have to double check.

@bpasero most of the feedback in this review I would need regarding the first commit which makes the `uriLabelService` a core service. Could you please take a glance if this makes sense to you?

fyi @aeschli @sandy081

bpasero commented on src/vs/platform/uriLabel/common/uriLabel.ts:
```diff
@@ -41,18 +46,21 @@ export class UriLabelService implements IUriLabelService {
 	}
 
-	getLabel(resource: URI, relative: boolean, forceNoTildify?: boolean): string {
+	acquireContextService(contextService: IWorkspaceContextService): void {
```
@isidorn this is not something we should push because it violates our service injection model. possible solutions:
* extract functionality into a third service that both depend on
* implement the 2 services in one place (e.g. context service and configuration service do this)

sandy081 replied:
Configuration service is already have this issue because it is the first service that needs to be created. It cannot be taken as example and follow that pattern for other services.

I see the root cause of this issue is that context service depending on IURILabelService. Due to the above reason, we have to avoid Configuration Service depending on other services as much as possible. 

I think `IWorkspace` object should not have a name property and the label service should provide it. Whoever wants a label for workspace should ask IURILabelService. Removing the name from `IWorkspace` will remove the dependency on `IURILabelService`

isidorn commented:
@sandy081 thanks, that makes perfect sense, let me work on that and I will push changes here.

isidorn commented:
Based on the feedback I have removed the `IWorkspace.name` and now everybody hast to go to the `UriLabelService.getWorkspaceLabel` in order to get this.
This change reduced quite some complexity across the workbench.

@bpasero @sandy081 could you please review, most of the places it is just adapoting to the new way of getting the workspace name. A thing to keep in mind is the `ExHostWorkspace` which gets the name from the renderer side and stores it.

@sandy081 since we have now removed the `name` from the `IWorkspace` [this method](https://github.com/Microsoft/vscode/blob/a03b3b5066642ac0d2a6db8c7b3efe6f47ed83d8/src/vs/platform/workspace/common/workspace.ts#L81) is not 100% correct since it will also `return true` if I pass in the IWorkspaceIdentifier - not sure if that will break something but just so we are aware

isidorn commented:
@bpasero agree - I was thinking the same. Will do that on master.

isidorn commented:
@sandy081 gave a thumbs up while talking to him -> merging
